pimd/pim_cmd.c

```diff
@@ -5817,90 +5817,49 @@ DEFPY (interface_ip_pim_activeactive,
 	return pim_process_ip_pim_activeactive_cmd(vty, no);
 }
 
-DEFUN_HIDDEN (interface_ip_pim_ssm,
-	      interface_ip_pim_ssm_cmd,
-	      "ip pim ssm",
-	      IP_STR
-	      PIM_STR
-	      IFACE_PIM_STR)
-{
-	int ret;
-
-	ret = pim_process_ip_pim_cmd(vty);
-
-	if (ret != NB_OK)
-		return ret;
-
-	vty_out(vty,
-		"WARN: Enabled PIM SM on interface; configure PIM SSM range if needed\n");
-
-	return NB_OK;
-}
-
-DEFUN_HIDDEN (interface_ip_pim_sm,
-	      interface_ip_pim_sm_cmd,
-	      "ip pim sm",
-	      IP_STR
-	      PIM_STR
-	      IFACE_PIM_SM_STR)
-{
-	return pim_process_ip_pim_cmd(vty);
-}
 
 DEFPY (interface_ip_pim,
        interface_ip_pim_cmd,
-       "ip pim [passive$passive]",
+       "[no] ip pim [sm|ssm$ssm|dm$dm|sm-dm$smdm|passive$passive]",
+       NO_STR
        IP_STR
        PIM_STR
+       IFACE_PIM_SM_STR
+       IFACE_PIM_STR
+       IFACE_PIM_DM_STR
+       IFACE_PIM_SMDM_STR
        "Disable exchange of protocol packets\n")
 {
 	int ret;
+	enum pim_iface_mode mode;
 
-	ret = pim_process_ip_pim_cmd(vty);
-
-	if (ret != NB_OK)
-		return ret;
+	if (no) {
+		if (passive)
+			return pim_process_ip_pim_mode_cmd(vty, PIM_MODE_SPARSE);
 
-	if (passive)
-		return pim_process_ip_pim_passive_cmd(vty, true);
+		return pim_process_no_ip_pim_cmd(vty);
+	}
 
-	return CMD_SUCCESS;
-}
 
-DEFUN_HIDDEN (interface_no_ip_pim_ssm,
-	      interface_no_ip_pim_ssm_cmd,
-	      "no ip pim ssm",
-	      NO_STR
-	      IP_STR
-	      PIM_STR
-	      IFACE_PIM_STR)
-{
-	return pim_process_no_ip_pim_cmd(vty);
-}
+	ret = pim_process_ip_pim_cmd(vty);
 
-DEFUN_HIDDEN (interface_no_ip_pim_sm,
-	      interface_no_ip_pim_sm_cmd,
-	      "no ip pim sm",
-	      NO_STR
-	      IP_STR
-	      PIM_STR
-	      IFACE_PIM_SM_STR)
-{
-	return pim_process_no_ip_pim_cmd(vty);
-}
+	if (ret != NB_OK)
+		return ret;
 
-DEFPY (interface_no_ip_pim,
-       interface_no_ip_pim_cmd,
-       "no ip pim [passive$passive]",
-       NO_STR
-       IP_STR
-       PIM_STR
-       "Disable exchange of protocol packets\n")
-{
-	if (passive)
-		return pim_process_ip_pim_passive_cmd(vty, false);
+	if (dm)
+		mode = PIM_MODE_DENSE;
+	else if (smdm)
+		mode = PIM_MODE_SPARSE_DENSE;
+	else if (passive)
+		mode = PIM_MODE_PASSIVE;
+	else if (ssm) {
+		mode = PIM_MODE_PASSIVE;
+		vty_out(vty,
+			"WARN: Enabled PIM SM on interface; configure PIM SSM range if needed\n");
+	} else
+		mode = PIM_MODE_SPARSE;
```

Comment:
so we have 2 cut-n-paste functions... boo.

Reply:
if you are referring to `PIM_MODE_PASSIVE`, that was intentional actually, but I was discussing this the other day with Nathan. I may need to change some logic there.

Reply:
no the function is cut-n-pasted between the v4 and v6 versions of the file.  Why not have 1 function that both sides call?